Automattic/wp-calypso
# Framework: experiment with schemas to validate persisted redux data

### The Problem

If we persist redux data, what happens when our tree changes? Our data shape is different right? Currently we do nothing, which is a little like this:
![datashapes](https://cloud.githubusercontent.com/assets/1270189/12959140/f1bca4c0-cfe9-11e5-9a84-784f015b722a.gif)

But as a dev, lets say we run master, let the redux tree persist, and then switch to a branch that has some minor refactoring for an existing subtree.

We then see something like:
![trace](https://cloud.githubusercontent.com/assets/1270189/12959177/2ccaafc6-cfea-11e5-8087-8f184e7302eb.png)

Not good. A normal user can hit this case too, just by visiting our website, and the coming back 2 weeks later.

(Don't worry `persist-redux` is off in all envs!)

### Solution

Let's check to see if our data shapes match. If the persisted data doesn't match our schemas, we throw it out and rebuild it from scratch.

As a continuation from #2754, this is an example of how we might use schemas to validate persisted redux data. 

As a recap, in each reducer, `SERIALIZE` takes what's in memory and returns a simple JS object. Similarly in `DESERIALIZE` the reducer is given state that is a simple JS object. With this PR, we then validate if the data matches the shape that we expect. If it is valid, we further manipulate the JS object to match what the subtree expects in memory (this might be a no-op, or say instantiation of Immutable.js objects). If the data is invalid, we return the default initial state.

There is some developer overhead for maintaining schemas, but I find that it's a great way to understand what's in the tree (without spinning up Calypso), and it gives us a lot of flexibility in refactoring our tree, since if the shape doesn't match, we rebuild it from scratch.

The following uses state/sites/items as an example using json schema v4 and [tv4](https://github.com/geraintluff/tv4)

## Testing Instructions
* In config/development.json set "persist-redux" to true
* Navigate to "My Sites"
* Switch between a few sites
* Reload the page
* Sites should be loaded without errors
* Tests in client/state/sites should also pass
